packages/nuxt/src/core/runtime/nitro/handlers/error.ts

```diff
@@ -1,54 +1,44 @@
 
+import { isJsonRequest } from '../utils/error'
 import { useRuntimeConfig } from '#internal/nitro'
 import { useNitroApp } from '#internal/nitro/app'
-import { isJsonRequest, normalizeError } from '#internal/nitro/utils'
-
-export default <NitroErrorHandler> async function errorhandler (error: H3Error, event) {
-  // Parse and normalize error
-  const { stack, statusCode, statusMessage, message } = normalizeError(error)
+import type { NuxtPayload } from '#app/nuxt'
 
-  // Create an error object
-  const errorObject = {
-    url: event.path,
-    statusCode,
-    statusMessage,
-    message,
-    stack: import.meta.dev && statusCode !== 404
-      ? `<pre>${stack.map(i => `<span class="stack${i.internal ? ' internal' : ''}">${i.text}</span>`).join('\n')}</pre>`
-      : '',
-    // TODO: check and validate error.data for serialisation into query
-    data: error.data as any,
-  } satisfies Partial<NuxtPayload['error']> & { url: string }
-
-  // Console output
-  if (error.unhandled || error.fatal) {
-    const tags = [
-      '[nuxt]',
-      '[request error]',
-      error.unhandled && '[unhandled]',
-      error.fatal && '[fatal]',
-      Number(errorObject.statusCode) !== 200 && `[${errorObject.statusCode}]`,
-    ].filter(Boolean).join(' ')
-    console.error(tags, (error.message || error.toString() || 'internal server error') + '\n' + stack.map(l => '  ' + l.text).join('  \n'))
+export default <NitroErrorHandler> async function errorhandler (error, event, { defaultHandler }) {
+  if (isJsonRequest(event)) {
+    // let Nitro handle JSON errors
+    return
   }
+  // invoke default Nitro error handler (which will log appropriately if required)
+  const defaultError = await defaultHandler(error, event, { json: true })
 
   if (event.handled) { return }
 
-  // Set response code and message
-  setResponseStatus(event, (errorObject.statusCode !== 200 && errorObject.statusCode) as any as number || 500, errorObject.statusMessage)
+  // let Nitro handle redirect if appropriate
+  const statusCode = error.statusCode || 500
+  if (statusCode === 404 && defaultError.status === 302) {
+    return defaultError
+  }
 
-  // JSON response
-  if (isJsonRequest(event)) {
-    setResponseHeader(event, 'Content-Type', 'application/json')
-    return send(event, JSON.stringify(errorObject))
+  if (import.meta.dev && typeof defaultError.body !== 'string' && Array.isArray(defaultError.body.stack)) {
+    // normalize to string format expected by nuxt `error.vue`
+    defaultError.body.stack = defaultError.body.stack.join('\n')
   }
 
+  const errorObject = defaultError.body as Pick<NonNullable<NuxtPayload['error']>, 'error' | 'statusCode' | 'statusMessage' | 'message' | 'stack'> & { url: string, data: any }
+  errorObject.message ||= 'Server Error'
+
+  setResponseHeaders(event, {
+    'x-content-type-options': defaultError.headers['x-content-type-options'],
```

Comment:
We are discarding any possible future headers from default handler. could be `...defaultError.headers`

Reply:
I was wanting to avoid fragility in case you changed case of headers or added some additional ones that would break JS-enabled error pages (as we render in Nuxt).

Reply:
`Cache-Control` doesn't seem to be lowercase, fyi

Reply:
https://github.com/nitrojs/nitro/commit/e3866d04130b6ac9deb712a1f92515e85e880c43